### DISCUSSION TITLE
How to achieve DDR reset in the HSS program?

### DISCUSSION THREAD

**Discussion by gege0500** [polarfire-soc/polarfire-soc-documentation#314] [HSS and U-Boot]
How to achieve the LPDDR4_RESET_n pin output low level in the HSS program?
![ddr](https://user-images.githubusercontent.com/87520674/233330933-a290d76f-baab-49d3-b8c1-2b861152b028.jpeg)

**Comment by hughbreslin** [comment]
Hi @gege0500 are you trying to force a reset on the DDR part itself? I think setting this register bit should do it (but I don't have the ability to test it myself):

![image](https://user-images.githubusercontent.com/44862881/233406733-63bf1db2-a43d-49e6-84cf-587f5e82b63b.png)

Without more info on what you what to achieve its hard to help here...

**Reply by gege0500** [reply]
Thank you for your reply.
There is another issue: when the board is powered on, in the HSS program, if the DDR training fails, it will never pass. Even if the dog bites and resets, it cannot pass, but it may pass after being powered on again. If DDR training fails, why cannot hot start pass, but cold start can pass? We look forward to your reply.

**Reply by mallynch** [reply]
I would suspect the memory itself has changed state, probably through MR memory writes during initial training, and command address bus access is not possible to allow retraining, hence power cycle required. Is this your own board? It is possible to turn on log output of DDR training on the console, this will help diagnose.

**Reply by gege0500** [reply]
We have modified the delay function so that LPDDR4_RESET_n pin output low level reaches 600ms, which meets the reset requirements of the DDR chip. Why is it still unable to pass repeated training? It is reasonable to say that the chip has been fully reset after pulling down 600ms, and training will restart.We look forward to your reply.
![WechatIMG3](https://user-images.githubusercontent.com/87520674/233760821-35aae5da-7862-4c01-9ce6-476163e25127.jpeg)

**Reply by mallynch** [reply]
Hi, Agree, the above should reset the memory. To help get to the bottom of this can you open a support request with your FAE. They can direct your question. To short circuit process, they most likely will look for the portion of schematic relating to DDR, the MSS Configuration .cfg file and ddr boot logs, obtained by placing the following defines 
#define DEBUG_DDR_INIT
#define DEBUG_DDR_RD_RW_FAIL
 in the <Project-Name>/src/boards/<your-board>/platform_config/mpfs_hal_config/mss_config.h> file. This will log ddr training messages on MMUART0.
